clients/src/main/java/org/apache/kafka/common/utils/Bytes.java

```diff
@@ -0,0 +1,415 @@
+    }
+
+    /**
+     * Check if two byte arrays are lexicographically equal to each other.
+     *
+     * @param buffer1 left operand
+     * @param buffer2 right operand
+     * @return 0 if equal, &lt; 0 if left is less than right, etc.
+     */
+    public static boolean equals(final byte[] buffer1, final byte[] buffer2) {
+        // short circuit case, other shortcuts are in the other equals function
+        if (buffer1 == buffer2) return true;
+
+        return equals(buffer1, 0, buffer1.length, buffer2, 0, buffer2.length);
+    }
+
+    /**
+     * Check if two byte arrays are lexicographically equal to each other.
+     *
+     * @param buffer1 left operand
+     * @param buffer2 right operand
+     * @param offset1 Where to start comparing in the left buffer
+     * @param offset2 Where to start comparing in the right buffer
+     * @param length1 How much to compare from the left buffer
+     * @param length2 How much to compare from the right buffer
+     * @return 0 if equal, &lt; 0 if left is less than right, etc.
+     */
+    public static boolean equals(final byte[] buffer1, int offset1, int length1,
+                                 final byte[] buffer2, int offset2, int length2) {
+        // short circuit case
+        if (buffer1 == null || buffer2 == null)
+            return false;
+        if (buffer1 == buffer2 &&
+                offset1 == offset2 &&
+                length1 == length2)
+            return true;
+        if (length1 != length2)
+            return false;
+        if (length1 == 0)
+            return true;
+
+        // since we're often comparing adjacent sorted data,
+        // it's usual to have equal arrays except for the very last byte
+        // so check that first
+        if (buffer1[offset1 + length1 - 1] != buffer2[offset2 + length2 - 1]) return false;
+
+        return ((ByteArrayComparator) BYTES_LEXICO_COMPARATOR).compare(buffer1, offset1, length1, buffer2, offset2, length2) == 0;
+    }
+
+    /**
+     * Returns a new byte array, copied from the given {@code buf},
+     * from the position (inclusive) to the limit (exclusive).
+     * The position and the other index parameters are not changed.
+     *
+     * @param buf a byte buffer
+     * @return the byte array
+     * @see #toBytes(ByteBuffer)
+     */
+    public static byte[] getBytes(ByteBuffer buf) {
+        byte [] result = new byte[buf.remaining()];
+        buf.get(result);
+        return result;
+    }
+
+    /**
+     * Returns a new byte array, copied from the given {@code buf},
+     * from the index 0 (inclusive) to the limit (exclusive),
+     * regardless of the current position.
+     * The position and the other index parameters are not changed.
+     *
+     * @param buf a byte buffer
+     * @return the byte array
+     * @see #getBytes(ByteBuffer)
+     */
+    public static byte[] toBytes(ByteBuffer buf) {
+        ByteBuffer dup = buf.duplicate();
+        dup.position(0);
```

Comment:
I haven't fully digested the way these things are being used, but this seems like an odd/fragile way to use ByteBuffers. Why not assume that the caller has arranged position/limit as appropriate, rather than force the copy to start at position(0)?